cli/crates/common/src/analytics.rs

```diff
@@ -0,0 +1,174 @@
+                    client: RudderAnalytics::load(write_key.to_owned(), dataplane_url.to_owned()),
+                    session_id: Ulid::new(),
+                    start_time: Utc::now(),
+                    version: env!("CARGO_PKG_VERSION").to_owned(),
+                }),
+            )
+            .expect("cannot set analytics twice");
+
+        Ok(())
+    }
+
+    fn write_data(data: &GrafbaseConfig) -> Result<(), CommonError> {
+        let data_location = Self::get_config_location();
+        if let Some(parent) = data_location.parent() {
+            std::fs::create_dir_all(parent).map_err(CommonError::CreateUserDotGrafbaseFolder)?;
+        }
+        std::fs::write(data_location, data.to_string()).map_err(CommonError::WriteAnalyticsDataFile)
+    }
+
+    fn to_value(&self) -> Value {
+        serde_json::to_value(self).expect("must parse")
+    }
+
+    #[must_use]
+    pub fn get_context(&self) -> Value {
+        self.to_value()
+    }
+
+    fn init_data() -> Result<(), CommonError> {
+        Self::write_data(&GrafbaseConfig {
+            enable_analytics: true,
+            anonymous_id: Some(Ulid::new()),
+        })
+    }
+
+    fn get_config_location() -> PathBuf {
+        Environment::get().user_dot_grafbase_path.join(GRAFBASE_CONFIG_FILE)
+    }
+
+    fn config_exists() -> Result<bool, CommonError> {
+        match Self::get_config_location().try_exists() {
+            Ok(result) => Ok(result),
+            Err(error) => Err(CommonError::ReadAnalyticsDataFile(error)),
+        }
+    }
+
+    fn read_data() -> Result<Option<GrafbaseConfig>, CommonError> {
+        if !Self::config_exists()? {
+            return Ok(None);
+        }
+
+        let data_path = Self::get_config_location();
+
+        let data_string = std::fs::read_to_string(data_path).map_err(CommonError::ReadAnalyticsDataFile)?;
+
+        let data: GrafbaseConfig =
+            serde_json::from_str(&data_string).map_err(|_| CommonError::CorruptAnalyticsDataFile)?;
+
+        Ok(Some(data))
+    }
+
+    /// # Panics
+    ///
+    /// panics if the static [`ANALYTICS`] object was not previously initialized
+    pub fn get() -> &'static Option<Self> {
+        match ANALYTICS.get() {
+            Some(analytics) => analytics,
+            // must be initialized in `main`
+            #[allow(clippy::panic)]
+            None => panic!("the analytics object is uninitialized"),
+        }
+    }
+
+    pub fn track(event_name: &str, properties: Option<Value>) {
+        let event_name = event_name.to_owned();
+        Self::get()
+            .as_ref()
+            .zip(Self::read_data().ok().flatten().and_then(|data| data.anonymous_id))
+            .map(|(analytics, anonymous_id)| {
+                // purposely ignoring errors
+                // TODO possibly change this to a long lived thread once we add more events
+                thread::spawn(move || {
```

Comment:
Should we store the join handle somewhere and block with it before exiting?

Reply:
We purposely don't since you won't want to make the CLI run longer due to an analytics request. If it succeeds it succeeds and if it doesn't it doesn't